lib/fake_idp/encryptor.rb

```diff
@@ -0,0 +1,89 @@
+require "xmlenc"
+
+module FakeIdp
+  class Encryptor
+    ENCRYPTION_STRATEGY = "aes256-cbc".freeze
+    KEY_TRANSPORT = "rsa-oaep-mgf1p".freeze
+
+    attr_reader :raw_xml, :certificate, :encryption_key
+
+    def initialize(raw_xml, certificate)
+      @raw_xml = raw_xml
+      @certificate = certificate
+    end
+
+    # Encryption approach borrowed from
+    # https://github.com/saml-idp/saml_idp/blob/master/lib/saml_idp/encryptor.rb
+    def encrypt
+      encryption_template = Nokogiri::XML::Document.parse(build_encryption_template).root
+      encrypted_data = Xmlenc::EncryptedData.new(encryption_template)
+      @encryption_key = encrypted_data.encrypt(raw_xml)
+      encrypted_key_node = encrypted_data.node.at_xpath(
+        "//xenc:EncryptedData/ds:KeyInfo/xenc:EncryptedKey",
+        Xmlenc::NAMESPACES,
+      )
+      encrypted_key = Xmlenc::EncryptedKey.new(encrypted_key_node)
+      encrypted_key.encrypt(openssl_cert.public_key, encryption_key)
+
+      xml = Builder::XmlMarkup.new
+      xml.EncryptedAssertion(xmlns: "urn:oasis:names:tc:SAML:2.0:assertion") do |enc_assert|
+        enc_assert << encrypted_data.node.to_s
+      end
+    end
+
+    private
+
+    def openssl_cert
+      @_openssl_cert ||= if certificate.is_a?(String)
+                           OpenSSL::X509::Certificate.new(Base64.decode64(certificate))
+                         else
+                           certificate
```

Comment:
What else could certificate be in this case?

Reply:
We want the certificate to be an instant of `OpenSSL::X509::Certificate`. If it's not a String it's likely already instantiated so we just return that. Otherwise, we instantiate it.